### ISSUE TITLE
Why is the output difference in gradient_check.py very large?

### ISSUE THREAD

**Issue by happyfire** [oreilly-japan/deep-learning-from-scratch#85]
All the code remains the same, and executing train_neuralnet.py with neural network layers can achieve good accuracy. However, when I run gradient_check.py to compare grad_numerical and grad_backprop, I am getting relatively large average differences:

W1: 0.0008748665632198089
b1: 0.004921597605313159
W2: 0.013747869077901162
b2: 0.27968619943647954
What could be the possible reasons for this? I have double-checked the code and found no discrepancies.

**Comment by happyfire** [comment]
I found the reason! It's because an old version of softmax is being used:
def softmax(a):  
    c = np.max(a)  
    exp_a = np.exp(a - c)  
    sum_exp_a = np.sum(exp_a)  
    y = exp_a / sum_exp_a  
    return y
    
    Although I have modified the source code to support a new version that handles 2D arrays, it seems that the cache has not been updated.
